storage/innobase/buf/buf0buf.cc

```diff
@@ -2845,42 +2839,43 @@ buf_page_get_low(
 			if (UNIV_UNLIKELY(e != DB_SUCCESS)) {
 				goto ibuf_merge_corrupted;
 			}
+		} else if (state < buf_page_t::UNFIXED) {
+			block->page.lock.x_unlock();
+			goto ignore_block;
 		}
 
-		if (rw_latch == RW_X_LATCH) {
-			goto get_latch_valid;
-		} else {
+#ifdef BTR_CUR_HASH_ADAPT
+		btr_search_drop_page_hash_index(block, true);
+#endif /* BTR_CUR_HASH_ADAPT */
+
+		switch (rw_latch) {
+		case RW_NO_LATCH:
 			block->page.lock.x_unlock();
-			goto get_latch;
+			break;
+		case RW_S_LATCH:
+			block->page.lock.x_unlock();
+			block->page.lock.s_lock();
+			break;
+		case RW_SX_LATCH:
+			block->page.lock.x_u_downgrade();
+			break;
+		default:
+			ut_ad(rw_latch == RW_X_LATCH);
 		}
+
+		mtr->memo_push(block, mtr_memo_type_t(rw_latch));
 	} else {
-get_latch:
 		switch (rw_latch) {
 		case RW_NO_LATCH:
 			mtr->memo_push(block, MTR_MEMO_BUF_FIX);
 			return block;
 		case RW_S_LATCH:
 			block->page.lock.s_lock();
-			ut_ad(!block->page.is_read_fixed());
-			if (UNIV_UNLIKELY(block->page.id() != page_id)) {
-				block->page.lock.s_unlock();
-				block->page.lock.x_lock();
-				goto page_id_mismatch;
-			}
-get_latch_valid:
```

Comment:
For my edification, why is the page ID mismatch check is now not needed in the "else" part of the "if" condition ?

```
 if (state >= buf_page_t::UNFIXED
            && allow_ibuf_merge
            && fil_page_get_type(block->page.frame) == FIL_PAGE_INDEX
            && page_is_leaf(block->page.frame)) {

```

Reply:
The function `buf_pool_t::corrupted_evict()` does two things before releasing the exclusive page latch: Invalidate the page ID, and set the `state` to `FREED` while preserving the buffer-fix count. In the `else` branch we detect the latter:
```c++
		if (UNIV_UNLIKELY(state < buf_page_t::UNFIXED)) {
			mtr->release_last_page();
			goto ignore_unfixed;
		}
```
In the `if` branch the condition for page ID mismatch could in fact be redundant; the `goto ignore_block` or `goto ignore_unfixed` for the `state` mismatch should do the same. It is the responsibility of `buf_page_t::read_complete()` to invoke `buf_pool_t::corrupted_evict()` when needed. I will revise the logic and remove the redundant check for page ID mismatch.